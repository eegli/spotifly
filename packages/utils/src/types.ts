export type Result<T> =
  | {
      success: true;
      value: T;
    }
  | {
      success: false;
      error: string;
    };
