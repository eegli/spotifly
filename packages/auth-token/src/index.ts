export { authorize } from './authorize';
export type { Options } from './types';
