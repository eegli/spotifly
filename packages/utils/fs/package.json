{
  "main": "dist/spotifly-utils-fs.cjs.js",
  "module": "dist/spotifly-utils-fs.esm.js"
}
