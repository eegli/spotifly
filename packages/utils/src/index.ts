export * as colors from './colors';
export { writeJSON } from './fs';
export type { WriteJSONOptions } from './fs';
export { chunkify } from './iterable';
