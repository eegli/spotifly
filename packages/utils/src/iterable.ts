/**
 * Splits an iterable into chunks of a specified size
 *
 * @param {Array<T>|Set<T>} obj - An iterable object
 * @param {number} chunkSize - A optional string param
 * @return {T[][]} The chunked iterable as an array
 *
 */
export function chunkify<T>(obj: T[] | Set<T>, chunkSize: number): T[][] {
  if (chunkSize < 1)
    throw new TypeError('Chunk size must be strictly positive');
  const len = Array.isArray(obj) ? obj.length : obj.size;
  const chunks = [];
  for (let i = 0; i < len; i += chunkSize) {
    chunks.push([...obj].slice(i, i + chunkSize));
  }
  return chunks;
}
