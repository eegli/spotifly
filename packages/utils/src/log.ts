import * as colors from './colors';

export default {
  error: (message: string) => {
    console.error(colors.red(message));
  },
  info: (message: string) => {
    console.info(colors.green(message));
  },
  log: (...messages: string[]) => {
    console.log(...messages);
  },
};
