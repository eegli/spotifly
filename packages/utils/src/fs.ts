import fs from 'fs-extra';
import { join } from 'path';

export type WriteJSONOptions = {
  path: string;
  fileName: string;
  data: unknown;
  compact?: boolean;
};

export const writeJSON = async ({
  path,
  fileName,
  data,
  compact = false,
}: WriteJSONOptions): Promise<string> => {
  if (!fileName.endsWith('.json')) {
    fileName += '.json';
  }
  path = join(process.cwd(), path);
  data = compact ? JSON.stringify(data) : JSON.stringify(data, null, 2);

  await fs.outputFile(join(path, fileName), data);
  return path;
};
