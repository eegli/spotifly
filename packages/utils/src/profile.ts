import { existsSync, readFileSync } from 'fs';
import ini from 'ini';
import os from 'os';
import { join } from 'path';
import type { Result } from './types';

const configFileName = '.spotifly';

type ProfileEntry = {
  clientId: string;
  clientSecret: string;
  refreshToken: string;
};

// Wrap in a helper object for testing
export const helpers = {
  readConfigWithPath(): [string, string] | null {
    let configPath = join(process.cwd(), configFileName);
    if (!existsSync(configPath)) {
      configPath = join(os.homedir(), configFileName);
    }
    if (existsSync(configPath)) {
      return [readFileSync(configPath, 'utf-8'), configPath];
    }
    return null;
  },
};

function collectProfiles(
  profiles: Record<string, unknown>,
): Map<string, ProfileEntry> {
  return Object.entries(profiles).reduce((acc, [key, value]) => {
    if (
      typeof value === 'object' &&
      value !== null &&
      'spt_client_id' in value &&
      'spt_client_secret' in value &&
      'spt_refresh_token' in value &&
      typeof value.spt_client_id === 'string' &&
      typeof value.spt_client_secret === 'string' &&
      typeof value.spt_refresh_token === 'string'
    ) {
      acc.set(key, {
        clientId: value.spt_client_id,
        clientSecret: value.spt_client_secret,
        refreshToken: value.spt_refresh_token,
      });
    }
    return acc;
  }, new Map<string, ProfileEntry>());
}

function readConfig(): Result<{
  profiles: Map<string, ProfileEntry>;
  configPath: string;
}> {
  const configWithPath = helpers.readConfigWithPath();
  if (!configWithPath) {
    return {
      success: false,
      error: 'Config file not found',
    };
  }
  const [config, configPath] = configWithPath;
  const profiles = collectProfiles(ini.parse(config));
  return {
    success: true,
    value: {
      profiles,
      configPath,
    },
  };
}

export function profilesFromConfig(): Result<[string[], string]> {
  const valueOrError = readConfig();
  if (!valueOrError.success) {
    return valueOrError;
  }
  const { configPath, profiles } = valueOrError.value;
  return {
    success: true,
    value: [[...profiles.keys()], configPath],
  };
}

export function credentialsFromConfig(profile: string): Result<ProfileEntry> {
  const valueOrError = readConfig();
  if (!valueOrError.success) {
    return valueOrError;
  }
  const profileEntry = valueOrError.value.profiles.get(profile);
  if (!profileEntry) {
    return {
      success: false,
      error: `Profile "${profile}" does not exist or is missing credentials`,
    };
  }
  return { success: true, value: profileEntry };
}
