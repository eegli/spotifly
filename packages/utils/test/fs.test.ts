import fs from 'fs-extra';
import { WriteJSONOptions } from '../src/fs';
import { writeJSON } from '../src/index';

jest.mock('fs-extra');
const mockFS = fs as jest.Mocked<typeof fs>;

jest.spyOn(process, 'cwd').mockImplementation(() => '/usr/dir');

const JSONspy = jest.spyOn(JSON, 'stringify');

beforeEach(() => {
  jest.clearAllMocks();
});

const data = {
  foo: 'bar',
  nested: {
    baz: 'lolilol',
  },
};

describe('FS, writeJSON', () => {
  const writeParams: WriteJSONOptions[] = [
    { path: '', fileName: 'test', data },
    { path: 'tmp', fileName: 'test', data },
    { path: 'tmp/dir', fileName: 'test.json', data },
    { path: '/tmp/dir', fileName: 'test.json', data },
    { path: '../dir', fileName: 'test.json', data },
  ];
  writeParams.forEach((args, idx) => {
    it(`writes data, ${idx}`, async () => {
      const path = await writeJSON(args);
      expect(mockFS.outputFile).toHaveBeenCalledTimes(1);
      expect(JSONspy).toHaveBeenCalledWith(data, null, 2);
      const cleanPath = path.replace(/\\|\//gi, '/');
      expect(cleanPath).toMatchSnapshot(`file ${idx}`);
    });
  });

  it('has compact option', async () => {
    await writeJSON({
      path: '',
      fileName: 'test',
      data,
      compact: true,
    });
    expect(JSONspy).toHaveBeenCalledWith(data);
  });
});
