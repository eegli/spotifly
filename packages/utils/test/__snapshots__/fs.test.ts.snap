// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fs writes data, 0: file 0 1`] = `"/usr/dir/test.json"`;

exports[`fs writes data, 1: file 1 1`] = `"/usr/dir/tmp/test.json"`;

exports[`fs writes data, 2: file 2 1`] = `"/usr/dir/tmp/dir/test.json"`;

exports[`fs writes data, 3: file 3 1`] = `"/usr/dir/tmp/dir/test.json"`;

exports[`fs writes data, 4: file 4 1`] = `"/usr/dir/test.json"`;
